### PR TITLE
feat: track combat task points unlock progress

### PR DESCRIPTION
**Pull request by iProdigy**
Tracks total points achieved and next unlock progress via varbits (and calculations)

image

### CODE REVIEW

File: src/main/java/dinkplugin/notifiers/data/CombatAchievementData.java [hunk 1]
```diff
@@ -4,23 +4,77 @@
...
+     */
+    @Nullable
+    Integer tierTotalPoints;
+
+    /**
+     * The tier whose rewards were just unlocked,
+     * <i>if</i> the player just completed the tier.
+     */
+    @Nullable
+    CombatAchievementTier justCompletedTier;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/data/CombatAchievementData.java:69]: Any ideas for a better variable name for the tier that was just completed upon rewards unlock?

**Review by Felanbird — APPROVED** [review]
image

File: src/main/java/dinkplugin/notifiers/CombatTaskNotifier.java [hunk 2]
```diff
@@ -41,19 +77,49 @@ private void handle(CombatAchievementTier tier, String task) {
...
+
+            boolean crossedThreshold = prevThreshold > 0 && totalPoints - taskPoints < prevThreshold;
+            CombatAchievementTier completedTier = crossedThreshold ? prev.getValue() : null;
+            String completedTierName = completedTier != null ? completedTier.getDisplayName() : "N/A";
+
+            String player = Utils.getPlayerName(client);
+            String message = StringUtils.replaceEach(
+                crossedThreshold ? config.combatTaskUnlockMessage() : config.combatTaskMessage(),
+                new String[] { "%USERNAME%", "%TIER%", "%TASK%", "%POINTS%", "%TOTAL_POINTS%", "%COMPLETED%" },
+                new String[] { player, tier.toString(), task, String.valueOf(taskPoints), String.valueOf(totalPoints), completedTierName }
```
Comment (pajlada) [inline src/main/java/dinkplugin/notifiers/CombatTaskNotifier.java:105]: I see we always include a replacement for `%COMPLETED%` - I am perfectly fine with this solution :+1:

File: src/main/java/dinkplugin/notifiers/data/CombatAchievementData.java (hunk 1 above)
Comment (pajlada) [reply src/main/java/dinkplugin/notifiers/data/CombatAchievementData.java:69]: This is verbose but it's clear what it does :+1:

**Review by pajlada — APPROVED** [review]
LGTM :+1: haven't tested in-game

**Pull request merged**